django/test/testcases.py

```diff
@@ -887,6 +887,57 @@ def assertXMLNotEqual(self, xml1, xml2, msg=None):
                 standardMsg = '%s == %s' % (safe_repr(xml1, True), safe_repr(xml2, True))
                 self.fail(self._formatMessage(msg, standardMsg))
 
+    def assertFormValid(self, form, msg=None):
+        """
+        Asserts that a form is valid
+        """
+        if not form.is_valid():
+            standardMsg = 'Form validation failed. Error(s):\n'
+            for field, field_errors in form.errors.items():
+                standardMsg += f'- {field}:\n'
+                for error in field_errors:
+                    standardMsg += f'    - {error}\n'
+            self.fail(self._formatMessage(msg, standardMsg))
+
+    def assertFormNotValid(self, form, expected_errors=None,
+                           allow_all_errors=True, msg=None):
+        """
+        Asserts that a form is not valid
+        """
+        error_list = []
+        if expected_errors:
+            for field, errors in expected_errors.items():
+                if isinstance(errors, tuple):
+                    for error in errors:
+                        error_list.append((field, error))
+                else:
+                    error_list.append((field, errors))
+
+        to_report = ''
+        if form.is_valid():
+            to_report = 'Form is valid\n'
+        else:
+            for field, error in error_list:
+                # if test passes additional expected errors which are not
+                # found, fail the test
+                if not form.has_error(field, error):
+                    to_report = f"('{field}', '{error}') expected error not found.\n"
+
+            if not allow_all_errors:
+                other_errors = []
+                for field, errors in form.errors.as_data().items():
+                    for error in errors:
+                        if not (field, error.code) in error_list:
+                            other_errors.append((field, error))
+                if other_errors:
+                    to_report += 'Other errors reported:\n'
+                    for field, error in other_errors:
+                        to_report += f'    - {field}: {error.messages[0]}\n'
```

Comment:
I think we should build an `ErrorDict` of expected errors and use `assertEqual()` to compare it with `form.error`. This way we will avoid building a custom message.

See also ticket-24782 that should be fixed in advance.

Reply:
@felixxm thanks for this, I'll have a look at `ErrorDict`. 

On your second comment I am missing something here (or did you link to the wrong ticket).

Reply:
Ahh, yes sorry wrong ticket. It should be ticket-28507.

Reply:
Now ticket-28507 is fixed I will go away and think about this more deeply. Currently, I am missing something, as I for `ValidationError` to pass we would need the user to pass in `code`, `message` and `params`. While this would allow for a more specific test I think this conflicts with simplicity. 

I'll have a think and to try and set out my proposal.